stdlib/socket/from.go

```diff
@@ -0,0 +1,309 @@
+	spec := new(FromSocketOpSpec)
+
+	if url, err := args.GetRequiredString("url"); err != nil {
+		return nil, err
+	} else {
+		spec.URL = url
+	}
+
+	if d, ok, err := args.GetString("decoder"); err != nil {
+		return nil, err
+	} else if ok {
+		spec.Decoder = d
+	} else {
+		spec.Decoder = "csv"
+	}
+
+	if !decoders[spec.Decoder] {
+		return nil, fmt.Errorf("unknow decoder %v: decoder must be one of %v", spec.Decoder, decoders)
+	}
+
+	return spec, nil
+}
+
+func newFromSocketOp() flux.OperationSpec {
+	return new(FromSocketOpSpec)
+}
+
+func (s *FromSocketOpSpec) Kind() flux.OperationKind {
+	return FromSocketKind
+}
+
+type FromSocketProcedureSpec struct {
+	plan.DefaultCost
+	URL     string
+	Decoder string
+}
+
+func newFromSocketProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*FromSocketOpSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+
+	return &FromSocketProcedureSpec{
+		URL:     spec.URL,
+		Decoder: spec.Decoder,
+	}, nil
+}
+
+func (s *FromSocketProcedureSpec) Kind() plan.ProcedureKind {
+	return FromSocketKind
+}
+
+func (s *FromSocketProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(FromSocketProcedureSpec)
+	ns.URL = s.URL
+	ns.Decoder = s.Decoder
+	return ns
+}
+
+func createFromSocketSource(s plan.ProcedureSpec, dsid execute.DatasetID, a execute.Administration) (execute.Source, error) {
+	spec, ok := s.(*FromSocketProcedureSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", s)
+	}
+
+	url := spec.URL
+	conn, err := net.Dial("tcp", url)
+	if err != nil {
+		return nil, errors.Wrap(err, "error in creating socket source")
+	}
+
+	return NewSocketSource(spec, conn, dsid, a)
+}
+
+func NewSocketSource(spec *FromSocketProcedureSpec, rc io.ReadCloser, dsid execute.DatasetID, a execute.Administration) (execute.Source, error) {
+	var decoder flux.ResultDecoder
+	switch spec.Decoder {
+	case "csv":
+		decoder = csv.NewResultDecoder(csv.ResultDecoderConfig{})
+	case "raw":
+		decoder = &RawDecoder{a: a}
+	}
+
+	if decoder == nil {
+		return nil, fmt.Errorf("unknown decoder type: %v", spec.Decoder)
+	}
+
+	itr := &socketIterator{
+		rc:      rc,
+		decoder: decoder,
+		data:    make(chan flux.Table),
+		err:     make(chan error),
+		done:    make(chan struct{}),
+	}
+	return execute.CreateSourceFromDecoder(itr, dsid, a)
+}
+
+type socketIterator struct {
+	rc      io.ReadCloser
+	decoder flux.ResultDecoder
+
+	data chan flux.Table
+	err  chan error
+	done chan struct{}
+}
+
+func (i *socketIterator) Connect() error {
+	result, err := i.decoder.Decode(i.rc)
+	if err != nil {
+		err = errors.Wrap(err, "decode error")
+	} else {
+		go func() {
+			err = result.Tables().Do(func(table flux.Table) error {
+				i.data <- table
+				return nil
+			})
+			if err != nil {
+				i.err <- err
+			}
+			close(i.data)
+			close(i.err)
+			close(i.done)
+		}()
```

Comment:
You should not need to use  a go routine inside of the SourceDecoder interface. This is a sign that the bare execute.Source interface should be used instead.

Reply:
Yeah, I see your point. I wanted to implement that interface because conceptually it is what I need, but it doesn't match the behavior of our decoders. Fixing to execute.Source would also fix the race problem.